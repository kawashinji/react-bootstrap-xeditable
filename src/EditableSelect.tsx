import React, { ChangeEvent } from 'react';
import XEditable from './XEditable';

export interface IOption {
  label?: string;
  text: string;
  value: string | number;
  disabled?: boolean;
}

export interface IEditableSelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  isEditing?: boolean;
  isEditable?: boolean;
  linkClassName?: string;
  nonValueText?: string;
  value: string | string[];
  onSave: (value: string | string[] | number) => void;
  onCancel?: () => void;
  options: string[] | IOption[];
}

interface IState {
  isEditing: boolean;
  options: IOption[];
  selectedOptions: IOption[];
}

export default class EditableTextArea extends React.Component<IEditableSelectProps, IState> {
  static defaultProps: Partial<IEditableSelectProps> = {
    isEditable: true,
    multiple: false
  };

  state: IState = {
    isEditing: !!this.props.isEditing,
    options: this.getOptions(),
    selectedOptions: this.findSelectedOptions(this.props.value) || []
  };

  getOptions(): IOption[] {
    return this.props.options && this.props.options[0] && (this.props.options[0] as IOption).text
      ? (this.props.options as IOption[])
      : (this.props.options as string[]).map((opt: string) => ({
          text: opt,
          value: opt
        }));
  }

  findSelectedOptions(value: string | string[] | number) {
    const options = this.getOptions();
    if (this.props.multiple) {
      const values = value as string[];
      const matchedOptions: IOption[] = [];
      for (const opt of options) {
        for (const val of values) {
          if (opt.value === val) {
            matchedOptions.push(opt);
          }
        }
      }

      return matchedOptions;
    }
    const filtered = options.filter((opt) => opt.value.toString() === value.toString());
    return filtered;
  }

  onChange = (event: ChangeEvent<HTMLSelectElement>) => {
    event.preventDefault();
    const selectedOptions = this.findSelectedOptions(event.currentTarget.value);
    this.setState({
      selectedOptions
    });
    if (this.props.onChange) {
      this.props.onChange(event);
    }
  };

  onSave = () => {
    this.setState({
      isEditing: false
    });
    if (this.props.multiple) {
      this.props.onSave(this.state.selectedOptions.map((opt: IOption) => opt.value.toString()));
    } else {
      this.props.onSave(this.state.selectedOptions[0].value);
    }
  };

  onCancel = () => {
    this.setState({
      isEditing: false,
      selectedOptions: this.findSelectedOptions(this.props.value)
    });
    if (this.props.onCancel) {
      this.props.onCancel();
    }
  };

  onLinkClick = () => {
    if (this.props.isEditable) {
      this.setState({
        isEditing: true
      });
    }
  };

  renderOptions() {
    return this.state.options.map((opt: IOption, index: number) => (
      <option key={`opt-${index}`} value={String(opt.value)} label={opt.label} disabled={opt.disabled}>
        {opt.text}
      </option>
    ));
  }

  renderEditing() {
    const { isEditing, isEditable, onSave, onCancel, ...otherProps } = this.props;
    const value =
      (this.state.selectedOptions.length > 0 &&
        (this.props.multiple
          ? this.state.selectedOptions.map((opt) => opt.value.toString())
          : this.state.selectedOptions[0].value)) ||
      '';
    return (
      <XEditable onSave={this.onSave} onCancel={this.onCancel}>
        <select {...otherProps} onChange={this.onChange} value={value}>
          {this.renderOptions()}
        </select>
      </XEditable>
    );
  }

  renderIcon() {
    if (!this.state.selectedOptions || this.state.selectedOptions.length === 0) {
      return <span className="glyphicon glyphicon-pencil" />;
    }
    return;
  }

  renderText() {
    if (this.state.selectedOptions.length !== 0) {
      if (this.props.multiple) {
        return this.state.selectedOptions.map((opt) => opt.text).join(',');
      }
      return this.state.selectedOptions[0].text;
    }
    return this.props.nonValueText;
  }

  render() {
    if (!this.props.isEditable) {
      return <div>{this.renderText()}</div>;
    }
    const classNames = ['editable', 'editable-click'];

    if (!this.state.selectedOptions || this.state.selectedOptions.length === 0) {
      classNames.push('editable-empty');
    }

    if (this.props.linkClassName) {
      classNames.concat(this.props.linkClassName.split(' '));
    }

    if (this.state.isEditing) {
      return this.renderEditing();
    } else {
      return (
        <a href="javascript:;" className={classNames.join(' ')} onClick={this.onLinkClick}>
          {this.renderText()}
          {this.renderIcon()}
        </a>
      );
    }
  }
}
