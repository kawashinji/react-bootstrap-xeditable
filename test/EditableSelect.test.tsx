import 'jest';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';
import renderer from 'react-test-renderer';
import EditableSelect from '../src/EditableSelect';

const options = [
  {
    text: 'China',
    value: 'CN'
  },
  {
    text: 'India',
    value: 'IN'
  }
];

it('renders correctly', () => {
  const tree = renderer.create(<EditableSelect name="test" options={options} value="IN" onSave={() => {}} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Render EditableSelect and Click it', () => {
  const wrapper = mount(<EditableSelect name="test" options={options} onSave={() => {}} />);
  wrapper.find('a').simulate('click');
  expect(toJSON(wrapper)).toMatchSnapshot();
});
