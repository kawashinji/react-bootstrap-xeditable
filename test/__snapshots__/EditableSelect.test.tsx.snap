// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render EditableSelect and Click it 1`] = `
<EditableTextArea
  isEditable={true}
  multiple={false}
  name="test"
  onSave={[Function]}
  options={
    Array [
      Object {
        "text": "China",
        "value": "CN",
      },
      Object {
        "text": "India",
        "value": "IN",
      },
    ]
  }
>
  <XEditable
    isLoading={false}
    onCancel={[Function]}
    onSave={[Function]}
  >
    <span
      className="editable-container editable-inline"
    >
      <div
        className="form-inline editableform"
      >
        <div
          className="control-group form-group"
        >
          <div>
            <div
              className="editable-input"
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <select
                multiple={false}
                name="test"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "text": "China",
                      "value": "CN",
                    },
                    Object {
                      "text": "India",
                      "value": "IN",
                    },
                  ]
                }
                value=""
              >
                <option
                  key="opt-0"
                  value="CN"
                >
                  China
                </option>
                <option
                  key="opt-1"
                  value="IN"
                >
                  India
                </option>
              </select>
            </div>
            <div
              className="editable-buttons"
            >
              <button
                className="btn btn-primary btn-sm editable-submit"
                onClick={[Function]}
                type="submit"
              >
                <i
                  className="glyphicon glyphicon-ok"
                />
              </button>
              <button
                className="btn btn-default btn-sm editable-cancel"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="glyphicon glyphicon-remove"
                />
              </button>
            </div>
          </div>
          <div
            className="editable-error-block help-block"
          />
        </div>
      </div>
    </span>
  </XEditable>
</EditableTextArea>
`;

exports[`renders correctly 1`] = `
<a
  className="editable editable-click"
  href="javascript:;"
  onClick={[Function]}
>
  India
</a>
`;
